src/zone.c

```diff
@@ -0,0 +1,191 @@
+/**
+ * collectd - src/zone.c
+ * Copyright (C) 2011       Mathijs Mohlmann
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; only version 2 of the License is applicable.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ *
+ * Authors:
+ *   Mathijs Mohlmann
+ **/
+
+#define _BSD_SOURCE
+
+#include "collectd.h"
+#include "common.h"
+#include "plugin.h"
+
+#include <sys/types.h>
+#include <sys/vm_usage.h>
+#include <procfs.h>
+#include <zone.h>
+
+#include "utils_avltree.h"
+
+#define	MAX_PROCFS_PATH	40
+#define MAX_ZONE_NAME 20
+#define FRC2PCT(pp)(((float)(pp))/0x8000*100)
+
+typedef struct zone_stats {
+	ushort_t      pctcpu;
+	ushort_t      pctmem;
+} zone_stats_t;
+
+static long pagesize;
+
+static int
+zone_compare(const zoneid_t *a, const zoneid_t *b)
+{
+	if (*a == *b)
+		return(0);
+	if (*a < *b)
+		return(-1);
+	return(1);
+}
+
+static int
+zone_read_procfile(char *pidstr, char *file, void *buf, size_t bufsize)
+{
+	int fd;
+
+	char procfile[MAX_PROCFS_PATH];
+	(void)snprintf(procfile, MAX_PROCFS_PATH, "/proc/%s/%s", pidstr, file);
+	while ((fd = open(procfile, O_RDONLY)) == -1) {
+		if ((errno != EMFILE) || (errno != ENFILE)) {
+			return(1);
+		}
+	}
+
+	if (pread(fd, buf, bufsize, 0) != bufsize) {
+		close(fd);
+		return (1);
+	}
+	close(fd);
+	return (0);
+}
+
+static int
+zone_submit_value(char *zone, gauge_t value)
+{
+	value_list_t vl = VALUE_LIST_INIT;
+	value_t      values[1];
+
+	values[0].gauge = value;
+
+	vl.values = values;
+	vl.values_len = 1; /*STATIC_ARRAY_SIZE (values);*/
+	sstrncpy (vl.host, hostname_g, sizeof (vl.host));
+	sstrncpy (vl.plugin, "zone", sizeof (vl.plugin));
+ 	sstrncpy (vl.type, "zone_cpu", sizeof (vl.type));
```

Comment:
I'm assuming that the `FRC2PCT()` macro results in a percentage. Please use the `percent` type here if that is so. Otherwise, please use the `vcpu` type.